csrc/scheduler/reduction_utils.cpp

```diff
@@ -736,18 +736,38 @@ class PersistentBufferProjector {
             (int)buffer_i, fusion_->inputs());
       }
     } else {
+      // visit consumer before producer. e.g.
+      // T1 = f(T0); Tx = add(T1, broadcast(sum(T1)));
+      // T2 = f(T1); Ty = add(T2, broadcast(sum(T2)));
+      // T3 = f(T2); Tz = add(T3, broadcast(sum(T3)));
+      // T1, T2, T3 are persistent buffers.
+      // The visiting order should be [T3, T2, T1].
+      // After project T3 to its producers, we have:
+      // Tz = add(f(T2),broadcast(sum(T3)));
+      // After project T2 to its producers, we have:
+      // Tz = add(f(f(T1)),broadcast(sum(T3)));
+      // Ty = add(f(T1), broadcast(sum(T2)));
+      // At last, the only persistent buffer is T1.
+      // For a solid case, see NVFuserTest.ChainProjectionToPersistentProducer.
+      std::vector<int> visiting_order(persistent_buffers.size());
+      std::iota(visiting_order.begin(), visiting_order.end(), 0);
+      std::sort(
```

Comment:
Couldn't this be a source of non-determinism? Shouldn't we use `std::stable_sort`?

Reply:
changed to `std::stable_sort`